app/middleware/acl.js

```diff
@@ -1,133 +1,116 @@
+    // and they don't have update permissions throw an error
+    if (!projectAccess(req.user, req.report)
+      || (UPDATE_METHODS.includes(req.method) && !(boundUser || hasMasterAccess(req.user)))
+    ) {
+      logger.error(`User: ${req.user.username} is trying to make a ${req.method} request to ${req.originalUrl}`);
+      return res.status(FORBIDDEN).json({
+        error: {message: 'You do not have the correct permissions to access this'},
+      });
     }
+  } else {
+    // See if route exists in special cases
+    const spCase = SPECIAL_CASES.find((value) => {
+      return route.match(value.path);
+    });
 
-
-    // Check if this is a report endpoint
-    if (this.req.report) {
-      // Check if this is a write endpoint and the user belongs
-      // to a group that is allowed to edit reports
-      if (['POST', 'PUT', 'DELETE'].includes(this.req.method)
-        && caseInsensitiveIntersect(userGroups, this.reportEdit).length > 0
-      ) {
-        // check if user is bound to report
-        const boundUser = this.req.report.users.some((reportUser) => {
-          return reportUser.user.ident === this.req.user.ident;
-        });
-
-        // They are allowed to edit if they belong to one of the groups in masterReportEdit
-        // or if they have been bound to the report
-        return caseInsensitiveIntersect(userGroups, this.masterReportEdit).length > 0 || boundUser;
-      }
-
-      // If read is not set to allow all, run check for read access
-      if (this.req.method === 'GET'
-        && (this.read.includes('*')
-        || caseInsensitiveIntersect(userGroups, this.read).length > 0)
+    // Check that route and method have special rules
+    if (spCase && spCase[req.method]) {
+      if (spCase[req.method].includes('*')
+        || intersectionBy(req.user.groups, spCase[req.method], 'name').length
       ) {
-        return true;
+        return next();
       }
 
-      return false;
-    }
-
-
-    // Non-report endpoints
-
-
-    // If the method is GET and the user has read access
-    // and/or everyone is allowed to read at this endpoint
-    if (this.req.method === 'GET'
-      && (this.read.includes('*')
-      || caseInsensitiveIntersect(userGroups, this.read).length > 0)
-    ) {
-      return true;
+      logger.error(`User: ${req.user.username} is trying to make a ${req.method} request to ${req.originalUrl}`);
+      return res.status(FORBIDDEN).json({
+        error: {message: 'You do not have the correct permissions to access this'},
+      });
     }
 
-    // If the method is an edit and the user has write permissions
-    // and/or everyone is allowed to write at this endpoint
-    if (['POST', 'PUT', 'DELETE'].includes(this.req.method)
-      && (this.write.includes('*')
-      || caseInsensitiveIntersect(userGroups, this.write).length > 0)
-    ) {
-      return true;
+    if ((UPDATE_METHODS.includes(req.method) && !hasMasterAccess(req.user))) {
```

Comment:
If I'm following the logic here correctly then this and the following lines assert that if a user has master access and is accessing a non-report route (and isn't a special case), then they should be able to proceed. Is that correct?

Reply:
This if statement is just saying that if this is not a special case and a user is trying to update something (POST, PUT, DELETE) and they don't have master access an error will be throw. Every authenticated user will be able to GET from the endpoints and only master access users can edit the data. The logic is staying the same as what was previously implemented. This ticket is just to move the ACL to middleware. More routes will be covered by these general settings now though.